Implementation of bip-0353

IgnacioPorte opened the pull request:
## What this does

Implements BIP-353 self-resolution: given `user@domain`, lndk fetches the DNSSEC-signed TXT record, validates the signatures locally, extracts the BOLT 12 offer, and pays it through the existing pay flow.

New gRPC method `PayHumanReadableAddress` and CLI `lndk-cli pay-hrn`.

DNS, DNSSEC validation and URI parsing are delegated to the [`bitcoin-payment-instructions`](https://crates.io/crates/bitcoin-payment-instructions)

## Not in this PR

BLIP-32 (getting the DNSSEC proof from a resolver peer over onion messages) is deferred, as discussed with @a-mpch. The LDK hook for that is `DNSResolverMessageHandler::handle_dnssec_query`.

## Testing                                                                                              
                                                                                                          
Unit tests and a mock-resolver itest cover the pay flow without real DNS.                               
                                                                                                          
For an end-to-end check:                                                                                
                                                                                                          
1. Set up three lnd+lndk nodes on regtest: payer, middle-hop, receiver.                                 
2. Open channels payer-middle and middle-receiver.                                                      
3. Create an offer on the receiver with `lndk-cli create-offer`.                                        
4. Publish it as a TXT record at `<user>.user._bitcoin-payment.<domain>` on a DNSSEC-signed zone.       
5. From the payer: `lndk-cli pay-hrn <user>@<domain>`.

a-mpch reviewed (commented):
solid start, let me know about the test related comments

a-mpch commented on src/dns_resolver.rs:
```diff
@@ -0,0 +1,89 @@
+        }
+    }
+
+    pub async fn resolve_name_to_offer(&self, name_str: &str) -> Result<String, OfferError> {
```
Nit: resolve_human_readable_name_to_offer 
Or resolver_hrn_to_offer

a-mpch commented on src/offers/handler.rs:
```diff
@@ -150,6 +164,30 @@ impl OfferHandler {
     }
 
+    /// Resolves a human-readable name (BIP-353) to an offer and pays it.
+    pub async fn pay_name(&self, cfg: PayNameParams) -> Result<Payment, OfferError> {
```
pay_hrn

a-mpch reviewed (commented):
Looks good! Haven't tested It yet.
Left some minor comments and we should be good to go

f3r10 reviewed (commented):
As indicated, I used `nacho@nachoporte.space` and the resolution was successful.

I tried to use other names, such:  `send.some@satsto.me` or `matt@mattcorallo.com`, but the resolution fails.

IgnacioPorte commented on src/dns_resolver.rs:
```diff
@@ -0,0 +1,89 @@
+
+        let resolution = self
+            .resolver
+            .resolve_hrn(&hrn_parsed)
```
No, I can add it

orbitalturtle commented on src/dns_resolver.rs:
```diff
@@ -0,0 +1,81 @@
+use bitcoin_payment_instructions::http_resolver::HTTPHrnResolver;
+use lightning::onion_message::dns_resolution::OMNameResolver;
+
+pub struct LndkDNSResolverMessageHandler {
```
nit: Lndk preface can be dropped?

orbitalturtle commented on src/lib.rs:
```diff
@@ -1,4 +1,5 @@
 mod clock;
+pub mod dns_resolver;
```
Reason for making this pub?

IgnacioPorte replied:
I originally made it pub because I didn't know a cleaner way to inject a mock from the itest. Turns out the rusty move is just depending on the trait, module is now plain mod dns_resolver and the itest injects via the new OfferHandler::with_hrn_resolver. Thanks for pushing on it.

orbitalturtle commented on src/dns_resolver.rs:
```diff
@@ -0,0 +1,111 @@
+        }
+    }
+
+    pub async fn resolver_hrn_to_offer(&self, name_str: &str) -> Result<String, OfferError> {
```
resolve_hrn_to_offer?

orbitalturtle replied:
Also, from testing this, it's hard to tell whether the resolution is successful without any logs. Please add some, maybe something like this:

```
info!("Resolving human-readable name: {}", name_str);
let resolved_uri = self.resolve_locally(name_str.to_string()).await?;
debug!("Resolved {} to URI: {}", name_str, resolved_uri);

let offer = self.extract_offer_from_uri(&resolved_uri)?;
info!("Successfully resolved {} to offer: {}", name_str, offer);
```

orbitalturtle commented:
Hey @IgnacioPorte, thanks for working on this. Are you still interested in pushing this over the line? 

Looks like it just needs a rebase and if you can squash the commits into one that'd be helpful as well. I'll take a deeper look this week and give it a test run as well. 

PS. Don't worry about the security audit CI check, the problem there should be resolved in the latest version of master.

orbitalturtle commented on src/cli.rs:
```diff
@@ -144,6 +147,34 @@ enum Commands {
+        /// The human-readable name to resolve (e.g., "user@example.com").
+        name: String,
+
+        /// Amount the user would like to pay. If this isn't set, we'll assume the user is paying
```
Amount in satoshis? in bitcoin? we need to specify or this is confusing

IgnacioPorte replied:
I added some missing details in the other commands too

orbitalturtle commented on src/cli.rs:
```diff
@@ -144,6 +147,34 @@ enum Commands {
         fee_limit_percent: Option<u32>,
     },
+    /// PayHumanReadableAddress pays a BOLT 12 offer by resolving a human-readable name (BIP-353).
+    PayHumanReadableAddress {
```
Let's shorten this to PayHRN or PayHRA? Just for the sake of making the lndk-cli a bit easier to type out - `lndk-cli pay-human-readable-address` is a bit long :)

IgnacioPorte commented:
> Hey @IgnacioPorte, thanks for working on this. Are you still interested in pushing this over the line?
> 
> Looks like it just needs a rebase and if you can squash the commits into one that'd be helpful as well. I'll take a deeper look this week and give it a test run as well.
> 
> PS. Don't worry about the security audit CI check, the problem there should be resolved in the latest version of master.

Hi, yeah of course, I will update it in the weekend.

orbitalturtle commented on tests/integration_tests.rs:
```diff
@@ -1440,3 +1440,129 @@ async fn test_receive_payment_from_offer_with_multiple_blinded_paths() {
+        &'a self,
+        _url: &'a str,
+    ) -> bitcoin_payment_instructions::hrn_resolution::HrnResolutionFuture<'a> {
+        Box::pin(async { Err("LNURL resolution not supported in mock") })
```
Does `unimplemented!()` work here and below? Seems a bit cleaner.

orbitalturtle reviewed (commented):
A few comments below, but this is lookin' good! Super excited for this to cross the line. I was able to test this with my own node and made some successful payments to a @lexe.app address 🎊 

In addition to the comments below, please also:
- Squash the commits into one, since this is a pretty singular change.
- Update the PR description to accurately reflect what's in it, since a lot of of it seems like it was saved for a future PR.
- Upgrade bitcoin-payment-instructions to the latest version 0.7.0. When trying to pay `send.some@satsto.me` and `matt@mattcorallo.com` I also got a resolution error about multiple TXT records, which looks like it should be resolved with this PR: https://github.com/rust-bitcoin/bitcoin-payment-instructions/pull/17

IgnacioPorte committed:
feat: pay BOLT 12 offers via BIP-353 human-readable names

Resolves a name like `alice@example.com` to a BOLT 12 offer over DoH
and pays it. DNSSEC is validated locally, as required by BIP-353.
Self-resolution only for now; peer-based resolution via BLIP-32 will
follow. Exposes a `PayHumanReadableAddress` gRPC endpoint and a
`lndk-cli pay-hrn` command.
